### PR TITLE
Offline Mode: Migration

### PR DESCRIPTION
**Pull request by kean**
Improves the migration from the previous version. The app will now display the previous "Local changes" status for all old posts that had revisions that never got uploaded.

## Testing

**Test 1.1**

**Precondition**: set breakpoint on `/rest/v1.2/sites/*/posts/*?` and make all requests to create and update posts fail.

- Create a new draft
- Tap "Publish"
- **Verify** that the request was sent (and failed)
- Enable "Syncronous Publishing" and restart the app
- **Verify** that the post was displayed in the "Draft" tab with a "Local changes" status
- Open the post and publish it
- **Verify** that the post got published

**Test 1.2**

**Precondition**: set breakpoint on `/rest/v1.2/sites/*/posts/*?`

- Create a new draft
- Save changes and wait until it syncs (request succeeds)
- Open the draft again and make changes
- Save the changes (request fails)
- Enable "Syncronous Publishing" and restart the app
- **Verify** that the post was displayed in the "Draft" tab with a "Local changes" status
- Open the post and tap "back" and "Save Draft"
- **Verify** that the new spinner got displayed (together with "Local changes") and the post got updated

**Test 1.3**

- **Verify** that for the revisions created using the new system, the "local changes" status isn't displayed

## Regression Notes
1. Potential unintended areas of impact

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Services/PostCoordinator.swift [hunk 1]
```diff
@@ -1142,7 +1149,7 @@ extension Foundation.Notification.Name {
 
 extension PostCoordinator: Uploader {
     func resume() {
-        guard isSyncPublishingEnabled else {
+        guard !isSyncPublishingEnabled else {
```
Comment (kean) [inline WordPress/Classes/Services/PostCoordinator.swift:1152]: 👀

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/Models/AbstractPost.h [hunk 2]
```diff
@@ -14,6 +14,14 @@ typedef NS_ENUM(NSUInteger, AbstractPostRemoteStatus) {
...
     AbstractPostRemoteStatusPushingMedia, // Push Media
     AbstractPostRemoteStatusAutoSaved,       // Post remote auto-saved
+
+    // All the previous states were deprecated in 24.7 and are no longer used
+    // by the app. To get the status of the uploads, use `PostCoordinator`.
+
+    /// The default state of the newly created local revision.
+    AbstractPostRemoteStatusLocalRevision,
+    /// The user saved the revision, and it needs to be uploaded to a server.
+    AbstractPostRemoteStatusSyncNeeded
```
Comment (kean) [inline WordPress/Classes/Models/AbstractPost.h:24]: It's a better way than using `confirmedChangesHash` for the following reasons:

- Once you confirm the save, the app "naturally" replaces the legacy `remoteStatus` with a new value `AbstractPostRemoteStatusSyncNeeded`
- Using integers is more efficient
- The DB migration in 24.8 will contain only removals – simpler

File: WordPress/Classes/Services/PostCoordinator.swift [hunk 3]
```diff
@@ -321,7 +321,8 @@ class PostCoordinator: NSObject {
         assert(isSyncAllowed(for: revision.original()), "Sync is not supported for this post")
 
         if !revision.isSyncNeeded {
-            revision.isSyncNeeded = true
+            revision.remoteStatus = .syncNeeded
+            revision.confirmedChangesTimestamp = Date()
```
Comment (kean) [inline WordPress/Classes/Services/PostCoordinator.swift:325]: We're not using this `confirmedChangesTimestamp` yet, but it's an important piece of information.

**Review by momo-ozawa — APPROVED** [review]
Works as described!

**Pull request merged**